Use less blur for distant directional shadow splits

Calinou opened the pull request:
`master` version of https://github.com/godotengine/godot/pull/60186.

This makes the transition between shadow splits less noticeable, specially when the expensive Blend Splits property is disabled. In many scenarios, you no longer need to enable Blend Splits to get good directional shadow appearance.

~~From empirical benchmarking, this seems to have a small performance cost (average 160.6 FPS over 5 screenshots instead of 163.8 FPS). I'm not sure if this is introduced by benchmarking. It's not very scientific since I carried this out from the editor and its fluctuating frame time information.~~
**Edit:** Further testing didn't reveal any performance difference.

This closes https://github.com/godotengine/godot-proposals/issues/2729.

**Note:** Not cherry-pickable to the `3.x` branch, since this relies on the per-light shadow blur feature found in the new Vulkan renderer.

`3.x` equivalent of this PR (doesn't work with Blend Splits enabled yet): https://github.com/Calinou/godot/tree/directional-shadow-distant-split-lower-blur-3.x

## TODO

- [x] Don't hardcode the blur factors. Instead, calculate them based on the effective "share".
  - By default, the multipliers would be:
    - `1.0` for the first split.
    - `0.5` for the second split.
    - `0.2` for the third split.
    - `0.1` for the fourth (and last) split.
  - These multipliers should change automatically when the split offsets are adjusted in the DirectionalLight.
- [x] Adapt the mobile renderer shader code as well.

## Preview

***Note:** GI and ambient lighting was disabled to make the difference in shadow rendering more obvious.*

### Before

*Distant shadows have less detail. There is a noticeable "cut" between splits near the lamp posts in the middle of the image.*

![Old shadows](https://user-images.githubusercontent.com/180032/118413757-2f086000-b6a1-11eb-8307-7d4bd18437a4.png)

### After

*Distant shadows keep a similar level of detail. Despite Blend Splits being disabled, the cut between splits is now much less noticeable.*

![New shadows](https://user-images.githubusercontent.com/180032/118413754-2d3e9c80-b6a1-11eb-9816-010ef5d9dc27.png)

jcostello commented:
I was about to create a proposal for this. Is this and the blend splits at draft?

Calinou commented:
> I was about to create a proposal for this. Is this and the blend splits at draft?

Yes, this PR needs to be finished before it can be merged.

Calinou committed:
Use less blur for distant directional shadow splits

This makes the transition between shadow splits less noticeable,
specially when the expensive Blend Splits property is disabled.

mrjustaguy commented:
Any reason why this PR has been dormant for so long, or just faded into the backlog of PRs to check?

clayjohn commented on servers/rendering/renderer_rd/shaders/scene_forward_clustered.glsl:
```diff
@@ -1378,62 +1378,78 @@ void main() {
 
 						pssm_coord = (directional_lights.data[i].shadow_matrix2 * v);
+						// Multiply shadow blur by the 3rd split's share to reduce discrepancy between shadow splits.
+						blur_factor = directional_lights.data[i].shadow_split_offsets.z / directional_lights.data[i].shadow_split_offsets.w;
```
Why is the blur for the second shadow split calculated based on the 3rd split?

Shouldn't this be based on the relative size of the split itself?

Calinou replied:
I don't really know why it's working this way, but it seems to be working correctly…

clayjohn replied:
I have a feeling that it is working by chance and that different settings won't produce the same nice results. We should come up with a more principled factor to reduce the amount of blur.

mrjustaguy replied:
The assumption of each split having half the resolution of the previous split is false, for a few reasons:
1) Each split has same Shadow map Resolution assigned to them (1/Split count of the total Atlas)
2) if Split 1 is 10% of the distance, and Split 2 is 30% of the max distance, the 2nd split isn't half the effective resolution of Split 1, as it's rendering 30%, including the 10% of Split 1, making it 1/3 of the effective resolution of Split 1.

Do keep in mind, the 10%,20%,50% splits are only the defaults.

Now that I think of it, there is possibly an easy way to Determine split blur probably...
When looking at the Default of 1 , 0.5 , 0.2 , 0.1 I notice a pattern. Default splits are 0.1 , 0.2 , 0.5 , 1
See where I'm going?
They're just in reverse order, which makes sense, as the effective shadow resolution is proportional to their size, and blur is connected to effective resolution so...

Calinou replied:
I should try playing around with split values in DirectionalLight3D and see if blurring still looks good.

clayjohn replied:
Right, so we are back to calculating the effective area based on split size. I guess we have two options then. We can adjust the blur amount with reference to the entire atlas, or with reference to the first split. 

I think adjusting with reference to the first split makes more sense (which is what this PR already does).

So, what we need to do is calculate the relative sizes one way of doing that is:
``shadow_split_offset.x / shadow_split_offset[i]``
With default values you get the same results as what you have (first split can still be hardcoded to 1.0). But with this, when you move away from default values the blur scale is determined with reference to the relative split size between the first split and the current split. Instead of being determined with reference to 1.0 and a different split.
*Before*
| split | size | blur_factor |
|-------|------|-------------|
| 1     | 0.1  | 1.0         |
| 2     | 0.2  | 0.5         |
| 3     | 0.5  | 0.2         |
| 4     |      | 0.1         |

| split | size | blur_factor |
|-------|------|-------------|
| 1     | 0.19  | 1.0         |
| 2     | 0.2  | 0.5         |
| 3     | 0.5  | 0.2         |
| 4     |      | 0.19         |

*After*
| split | size | blur_factor |
|-------|------|-------------|
| 1     | 0.1  | 1.0         |
| 2     | 0.2  | 0.5         |
| 3     | 0.5  | 0.2         |
| 4     |      | 0.1         |

| split | size | blur_factor |
|-------|------|-------------|
| 1     | 0.19  | 1.0         |
| 2     | 0.2  | 0.95         |
| 3     | 0.5  | 0.2         |
| 4     |      | 0.1         |

Calinou commented:
I applied @clayjohn's suggestions and tested again on both clustered and mobile backends. It looks good.

mrjustaguy commented:
Blend Splits with consistent blur levels alone probably becomes unnecessary for plenty of scenes. If Normal bias change adds to that.. especially if and when https://github.com/godotengine/godot-proposals/issues/3908 (spherical Splits) become a thing... Blend Splits will really be a nearly pointless feature for most use cases.. 😃

mrjustaguy commented:
Eh, I've noticed how Shadow Blur quality setting is now very inconsistent in what it's doing...

Calinou commented:
> For example, Soft Shadows Low look the softest, while going towards Ultra generally slowly decreases how soft they are, and the level of graininess is about the same.

I can reproduce this on the Vulkan Clustered backend, but not the Vulkan Mobile backend. Can you reproduce this on a vanilla `master` branch and previous 4.0 alphas?

clayjohn commented:
> For example, Soft Shadows Low look the softest, while going towards Ultra generally slowly decreases how soft they are, and the level of graininess is about the same.

This appears to be a regression in master. It looks like the enums are misaligned or something.

clayjohn reviewed (approved):
Looks great! Once we resolve that shadow regression in master this will look even better!

mrjustaguy commented:
Alpha 6 has no issues

mrjustaguy commented:
I think I may have figured out why there is such shadow acne appearing on a Plane when the plane is viewed from such an angle (85-90 degrees angle from Plane's Normal)
The Shadow Map for that Plane is turning into essentially a Pixel Line, which doesn't have enough information for Shadows to be Acne-free, and there is an extremely a Sharp change in the Depth.

There might be a solution for this, which would be to detect Large differences in Depth on the Shadow Map and based on that increase Bias for the areas of the Shadow Map that are affected..

Calinou commented:
Tested again on latest `master`:

### Before

![2022-04-11_19 46 48](https://user-images.githubusercontent.com/180032/162799379-f382b27c-6a67-440b-b746-acb8dc67f656.png)

### After

![2022-04-11_19 47 39](https://user-images.githubusercontent.com/180032/162799375-25b17511-11e4-4369-9f48-897446d93ecd.png)


Performance is pretty much identical.

**Testing project:** [test_shadow_b.zip](https://github.com/godotengine/godot/files/8466570/test_shadow_b.zip)

clayjohn reviewed (approved):
Yep, looks great rebased on master. Great work!

akien-mga commented:
Thanks!
